Remote API to connect additional code review tools

fo-code opened the pull request:
Adds a REST API that provides the coverage information of modified lines for each modified file - see #731 .

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/ChangedLinesModel.java:
```diff
@@ -0,0 +1,57 @@
+    private Type type;
+
+    public ChangedLinesModel(final int start_line, final int end_line, final Type type) {
+        this.setStartLine(start_line);
```
We should not call other methods from a constructor that can be overwritten. This could unexpectedly change the constructor's behavior in subclass. Directly set the attributes and make them final. 
Hence, we do not need setters here,

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/ChangedLinesModel.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public boolean equals(final Object o) {
```
These should be added in the end of the file since the getters are more of interest than equals/hashcode

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/Type.java:
```diff
@@ -0,0 +1,7 @@
+package io.jenkins.plugins.coverage.metrics.model;
+
+public enum Type {
```
Type is to generic as class name in my opinion. The reference to coverage is missing.
Maybe LineCoverageType?
A look at the used [Coverage-Model](https://github.com/jenkinsci/coverage-model) might be also helpful to see what is already implemented and how the model classes are named.

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/Side.java:
```diff
@@ -0,0 +1,6 @@
+package io.jenkins.plugins.coverage.metrics.model;
+
+public enum Side {
```
I am not sure if this is really necessary for coverage since always the right side / new code is of interest.
I think this can be removed.

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/LineCoverageViewModel.java:
```diff
@@ -0,0 +1,76 @@
+        this.filesWithChangedLines = getFilesWithChangedLines(node);
+    }
+
+    public static List<FileWithChangedLinesCoverageModel> getFilesWithChangedLines(final Node node) {
```
The business logic should not be part of the data class.

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/LineCoverageViewModel.java:
```diff
@@ -0,0 +1,76 @@
+    private final List<FileWithChangedLinesCoverageModel> filesWithChangedLines;
+
+    public LineCoverageViewModel(final Node node) {
+        this.filesWithChangedLines = getFilesWithChangedLines(node);
```
The constructor can get the list of FileWithChangedLinesCoverageModel and directly initialize the attribute here so it can be final.

fo-code commented on plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/LineCoverageViewTest.java:
```diff
@@ -0,0 +1,75 @@
+/**
+ * TODO: JavaDoc me
+ */
+class LineCoverageViewTest extends AbstractModifiedFilesCoverageTest {
```
The name of this test class does not really describe what is tested. After the business logic is seperated from the data classes, a test for specificely this business class can be written.
Example: MyBusinessClass.java -> MyBusinessClassTest.java

fo-code commented on plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/LineCoverageViewTest.java:
```diff
@@ -0,0 +1,75 @@
+        var fileOne = new FileWithChangedLinesCoverageModel("test/example/Test1.java");
+        fileOne.setListOfChangedLines(fileOneLinesList);
+
+        assertThat(filesWithChangedLines, IsCollectionContaining.hasItem(fileOne));
```
AssertJ would be preferable here instead of Hamcrest.
The assertion here is also duplicated, the assertions above are enough.
Also, something about the equals methods must not be correct since the assertion matches even if there are only three elements in fileOne and the first one combines two of the tested filesWithChangedLines.

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/FileWithChangedLinesCoverageModel.java:
```diff
@@ -0,0 +1,55 @@
+        }
+        FileWithChangedLinesCoverageModel file = (FileWithChangedLinesCoverageModel) o;
+        return Objects.equals(getRelativePath(), file.getRelativePath())
+                //&& Objects.equals(getListOfChangedLines(), file.getListOfChangedLines());
```
Why?

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/FileWithChangedLinesCoverageModel.java:
```diff
@@ -0,0 +1,55 @@
+
+@ExportedBean
+public class FileWithChangedLinesCoverageModel {
+    private String relativePath;
```
These should be private final fields set be the constructor.

fo-code reviewed (commented):
Good initial implementation - I added a few comments.
Please also have a look the the GitHub Checks and fix the Checkstyle, PMD, ... Warnings.

Generally, the model classes should not contain "Model" in their name in my opinion. Also, I think we can find better names for the classes to better match their use case - we should rethink that. "Changed Lines" here for example is always described as "Modified Lines". We should keep this consistent.

fo-code commented on plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtilTest.java:
```diff
@@ -13,10 +13,10 @@
- * Test to assert that all modified lines and their respective coverage types are correctly extracted from the
- * coverage tree created by the createCoverageTree() method, using the getFilesWithModifiedLines() method.
+ * Test to assert that all modified lines and their respective coverage types are correctly extracted from the coverage
+ * tree created by the createCoverageTree() method, using the calculateFilesWithModifiedLines() method in {@link CoverageApiUtil}.
```
Here the method can be linked directly

aelfric-m replied:
Done.

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtil.java:
```diff
@@ -10,57 +10,64 @@
     }
 
     /**
      * Static method to extract modified lines of code and their respective coverage status from passed Node object.
```
"Static method" is unnecessary information

aelfric-m replied:
Removed.

fo-code commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/LineCoverageType.java:
```diff
@@ -0,0 +1,10 @@
+public enum LineCoverageType {
+    COVERED,
+    MISSED,
+    PARTRIALLY_COVERED
```
Typo

aelfric-m replied:
Done

fo-code reviewed (commented):
Looks great - only a few comments.

Another idea: 
we can extend the FileWithModifiedLines with indirect coverage changes aswell (see FileNode#getIndirectCoverageChanges();).
This shows which lines have coverage changes caused by changed test cases. Would be also helpful i guess.
This requires another LineCoverageType: INDIRECT_CHANGE. Should be easy to include since the calculation is the same as for the other types.

fo-code commented:
@uhafner Nonetheless there are still a few tasks to do, the API now is in a final state.
What is your opinion on this? I would like to wait with merging until we have your confirmation. :)

aelfric-m commented on plugin/src/test/java/io/jenkins/plugins/coverage/metrics/AbstractModifiedFilesCoverageTest.java:
```diff
@@ -90,20 +92,23 @@ protected Node createCoverageTree() {
 
     /**
      * Gets the name of the test file without modified lines.
+     * @return TEST_FILE_NOT_MODIFIED returns the name of a file with no modified code lines.
```
Done.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtil.java:
```diff
@@ -10,57 +10,64 @@
+     *         type of coverage pertaining to each line of code (COVERED, MISSED, or PARTIALLY_COVERED)
      */
-    public static void getModifiedLineBlocks(final List<Integer> modifiedLines,
+    public static void calculateModifiedLineBlocks(final List<Integer> modifiedLines,
```
Done. Also corrected a type-o in PARTIALLY_COVERED.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtil.java:
```diff
@@ -10,57 +10,64 @@
 
-//TODO: rename to util
 /**
  * Server side model that provides data for the details of line coverage results in modified lines.
```
corrected.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtil.java:
```diff
@@ -10,57 +10,64 @@
-            var listOfMissedLines = new ArrayList<>(fileNode.getMissedLines());
-            var listOfPartialLines = new ArrayList<>(fileNode.getPartiallyCoveredLines().keySet());
-            List<Integer> listOfCoveredLines = new ArrayList<>();
+        for (FileNode fileNode : node.filterByModifiedLines().getAllFileNodes()) {
```
Done.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtil.java:
```diff
@@ -10,57 +10,64 @@
             int i = 0;
-            for (Integer a:fileNode.getLinesWithCoverage()) {
+            for (Integer a : fileNode.getLinesWithCoverage()) {
                 if (fileNode.getCoveredCounters()[i] > 0 && fileNode.getMissedCounters()[i] == 0) {
```
Corrected. Changed variable name to "line" and removed the counter variable i.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/ModifiedLinesBlock.java:
```diff
@@ -6,8 +6,9 @@
- * line number and the type of coverage of the block. Each object is associated with a FileWithModifiedLines object.
+ * Model class containing data pertained to consecutive lines of modified code. Each object possesses a starting and
+ * ending line number and the type of coverage of the block. Each object is associated with a FileWithModifiedLines
+ * object.
```
Done.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/FileWithModifiedLines.java:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * Model class containing data pertaining to files with modified lines of code.
+ * Each object contains a relative file path and a list of ModifiedLinesBlock objects.
```
Done.

aelfric-m commented on plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtilTest.java:
```diff
@@ -0,0 +1,62 @@
+        var checkTrue = fileOne.equals(fileOne);
+        var checkFalse = fileOne.equals(fileTwo);
+
+        assertThat(checkTrue).isTrue();
```
Done.

aelfric-m commented on plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtilTest.java:
```diff
@@ -0,0 +1,62 @@
+
+        var fileOne = new FileWithModifiedLines("test/example/Test1.java", fileOneLinesList);
+        var fileTwo = new FileWithModifiedLines("fileTwo", null);
+        var checkTrue = fileOne.equals(fileOne);
```
If I remove this, the line of code in the equals() method that returns true if the same object is passed to it as parameter, will not be covered by tests. I agree, that this doesn't look good, so am open for suggestions.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/FileWithModifiedLines.java:
```diff
@@ -0,0 +1,56 @@
+        }
+        FileWithModifiedLines file = (FileWithModifiedLines) o;
+        return Objects.equals(this.getRelativePath(), file.getRelativePath())
+                && this.getListOfModifiedLines().containsAll(file.getListOfModifiedLines())
```
Done.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/ModifiedLinesBlock.java:
```diff
@@ -0,0 +1,68 @@
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof ModifiedLinesBlock)) {
+            return false;
+        }
```
Can you use the same convention for equals as all other classes? Typically `getClass` is used in equals.

aelfric-m replied:
Changed.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/FileWithModifiedLines.java:
```diff
@@ -0,0 +1,58 @@
+ * Model class containing data pertaining to files with modified lines of code. Each object contains a relative file
+ * path and a list of {@link ModifiedLinesBlock} objects.
+ */
+@ExportedBean
```
Are you planning to use these classes outside of the API?

Otherwise I would recommend to create a new package (restapi, remoting, api, ...) and move all classes there. Then we can move the *Api classes to the new package as well. 

This would allow to make all internal classes package private as well.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/FileWithModifiedLines.java:
```diff
@@ -0,0 +1,58 @@
+    private final String relativePath;
+    private final List<ModifiedLinesBlock> listOfModifiedLines;
+
+    /**
+     * Constructor for the FileWithModifiedLines class.
+     *
+     * @param relativePath
+     *         the relative path of the file
+     * @param listOfModifiedLines
+     *         all modified lines inside the file
+     */
```
Either make package private (if possible, see above) and remove the comment. Or use something like:
```suggestion
    /**
     * Creates a new instance of {@link FileWithModifiedLines}.
     *
     * @param relativePath
     *         the relative path of the file
     * @param listOfModifiedLines
     *         all modified lines inside the file
     */
```

aelfric-m replied:
Done. Files have been made package private if possible, otherwise the correct format is being used.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/LineCoverageType.java:
```diff
@@ -0,0 +1,10 @@
+package io.jenkins.plugins.coverage.metrics.model;
+
+/**
+ * Enum to define the type of coverage assigned to any ModifiedLinesBlock object.
```
```suggestion
 * Defines the type of line coverage assigned to any {@link ModifiedLinesBlock} object.
```

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/model/ModifiedLinesBlock.java:
```diff
@@ -0,0 +1,68 @@
+    private final int endLine;
+    private final LineCoverageType type;
+
+    /**
+     * Constructor for this class.
+     *
+     * @param startLine
+     *         number of the first line of code associated with this block.
+     * @param endLine
+     *         number of the last line of code associated with this block.
+     * @param type
+     *         type of coverage.
+     */
```
See above. You should create a template in your IDE for such comments

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildAction.java:
```diff
@@ -609,6 +609,7 @@ protected String getBuildResultBaseName() {
 
     @Override
     public CoverageViewModel getTarget() {
+
```
Please remove unrelated formatting changes

aelfric-m replied:
Removed.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApiUtil.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Server side model that provides data for the details of line coverage results of modified lines.
+ */
+public class CoverageApiUtil {
```
This class and methods should be actually part of the API class. 

Then the constructor of `LineCoverageApi` works on a `Node`. You can make the method package private to test it there.

aelfric-m replied:
We have been on the fence about this. The logic is now part of the API class again.

uhafner reviewed (changes requested):
Thanks for your pull request! 

I will finish the review later this week (tests).

aelfric-m commented on plugin/pom.xml:
```diff
@@ -202,6 +202,12 @@
     </dependency>
 
     <!-- Test Dependencies -->
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <version>3.23.1</version>
+      <scope>test</scope>
+    </dependency>
```
Removed.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageApi.java:
```diff
@@ -136,7 +136,10 @@ public Result getOverallResult() {
         @Exported(inline = true)
         public Collection<QualityGateItemApi> getResultItems() {
-            return qualityGateResult.getResultItems().stream().map(QualityGateItemApi::new).collect(Collectors.toList());
+            return qualityGateResult.getResultItems()
+                    .stream()
+                    .map(QualityGateItemApi::new)
+                    .collect(Collectors.toList());
```
Removed.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageViewModel.java:
```diff
@@ -73,7 +73,8 @@ public class CoverageViewModel extends DefaultAsyncTableContentProvider implemen
 
     private static final ElementFormatter FORMATTER = new ElementFormatter();
-    private static final Set<Metric> TREE_METRICS = Set.of(Metric.LINE, Metric.INSTRUCTION, Metric.BRANCH, Metric.MUTATION);
+    private static final Set<Metric> TREE_METRICS = Set.of(Metric.LINE, Metric.INSTRUCTION, Metric.BRANCH,
+            Metric.MUTATION);
```
Removed.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageViewModel.java:
```diff
@@ -347,7 +357,7 @@ public String getSourceCode(final String fileHash, final String tableId) {
             try {
                 Node fileNode = targetResult.get();
-                return readSourceCode((FileNode)fileNode, tableId);
+                return readSourceCode((FileNode) fileNode, tableId);
```
Removed.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/CoverageApi.java:
```diff
@@ -1,4 +1,4 @@
-package io.jenkins.plugins.coverage.metrics.steps;
+package io.jenkins.plugins.coverage.metrics.restapi;
```
You need to add that new package to https://github.com/jenkinsci/code-coverage-api-plugin/blob/master/plugin/src/test/resources/design.puml to get the architecture tests running

aelfric-m replied:
Done.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/metrics/restapi/_api.jelly:
```diff
@@ -0,0 +1,17 @@
+<?jelly escape-by-default='true'?>
```
I'm not sure if this is the right package for that file? I can't see the content. Do you see that text?

fo-code replied:
Done. Has been in the wrong directory indeed.

aelfric-m commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLineCoverageApi.java:
```diff
@@ -0,0 +1,101 @@
+ * Remote API to list the details of modified line coverage results.
+ */
+@ExportedBean
+public class ModifiedLineCoverageApi {
```
This is true. The name still seems a bit clunky, but will do for now.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/gerrit/GerritCoverageController.java:
```diff
@@ -0,0 +1,111 @@
+import io.jenkins.plugins.util.JenkinsFacade;
+
+@Extension
+public class GerritCoverageController implements UnprotectedRootAction {
```
What is this class for? It does not belong to this plugin and should be moved somewhere else.

fo-code replied:
This is required for using the coverage API in Gerrit since there we have no information about the Jenkins build number that is required for accessing the API. 
We thought some other people could use it aswell, maybe on a different location where others could also add similar extension for other tools for which the API does not work out of the box? 

If this does not belong to this plugin in your opinion, we can only use it in our fork of course. :)

uhafner replied:
Yes, I would prefer to get that file removed. It would make sense that you put it in a small additional plugin that contains the Gerrit and Jenkins bridge.

fo-code commented:
@uhafner I extended the GitForensicsITest with a call to the new API to verify the resulting JSON. I added that in that test class since it offers a special setup where modified code lines as well as related coverage data is provided. Both is required for a decent test scenario here.
Also, I finished the implementation and added a description of the new API to the README.

From my side we can merge this PR, if you don't have any further suggestions.

uhafner commented on plugin/src/main/resources/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApi/_api.jelly:
```diff
@@ -0,0 +1,17 @@
+<?jelly escape-by-default='true'?>
```
This file needs to be moved to `[...]/coverage/metrics/restapi/CoverageApi/_api.jelly`. It should be shown when the user clicks on the API link on the `coverage` view.

fo-code replied:
Ah, I see. Thanks for the hint.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApi.java:
```diff
@@ -0,0 +1,97 @@
+                if (fileNode.getMissedOfLine(line) == 0 && fileNode.getCoveredOfLine(line) > 0) {
+                    coveredLines.add(line);
+                }
+            }
```
Wouldn't it make sense to move the method to the coverage model? (This can be done after the merge...)

fo-code replied:
Yes I think so :)

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApi.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @return a list of {@link FileWithModifiedLines} objects.
+     */
+    private static List<FileWithModifiedLines> createListOfFilesWithModifiedLines(final Node node) {
```
```suggestion
    private List<FileWithModifiedLines> createFilesWithModifiedLines(final Node node) {
```

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApi.java:
```diff
@@ -0,0 +1,97 @@
+
+            var modifiedLinesBlocks = new TreeSet<ModifiedLinesBlock>();
+
+            calculateModifiedLineBlocks(coveredLines, modifiedLinesBlocks, LineCoverageType.COVERED);
+            calculateModifiedLineBlocks(missedLines, modifiedLinesBlocks, LineCoverageType.MISSED);
+            calculateModifiedLineBlocks(partiallyCoveredLines, modifiedLinesBlocks, LineCoverageType.PARTIALLY_COVERED);
```
It is somehow untypical in Java to use a parameter as output value. Why don't you use a return value and add the result to the list?

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApi.java:
```diff
@@ -0,0 +1,97 @@
+     *         type of coverage pertaining to each line of code ({@link LineCoverageType#COVERED},
+     *         {@link LineCoverageType#MISSED}, or {@link LineCoverageType#PARTIALLY_COVERED})
+     */
+    private static void calculateModifiedLineBlocks(final List<Integer> modifiedLines,
```
```suggestion
    private void calculateModifiedLineBlocks(final List<Integer> modifiedLines,
```

uhafner commented on plugin/src/main/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApi.java:
```diff
@@ -0,0 +1,97 @@
+                }
+            }
+        }
+    }
```
I'm not sure if this is relevant: the result is not accurate as typically not all source code lines have a coverage counter (comments, empty lines). As a user I would assume that those not measured lines would not create a new lines block. See https://github.com/jenkinsci/coverage-model/blob/67df3455cb8577a872eecef5c8daae0564c445a7/src/main/java/edu/hm/hafner/coverage/FileNode.java#L576 for an alternative implementation of this problem.

fo-code replied:
Yes true, I implemented another solution that should fix this problem.

fo-code replied:
I improved the implementation like this:
- the API will provide only coverage information for modified lines (as described, has been wrong before)
- lines without coverage will be included in a block if:
  - the lines with coverage information before and after have the same coverage type
  - the lines without coverage information have also been modified (else, there is too much not required information since the user is only interested in modified lines here, in my opinion)
- I added Unit tests to cover this behavior

uhafner reviewed (changes requested):
Almost ready for merge, thanks for updating!

fo-code commented on plugin/src/test/java/io/jenkins/plugins/coverage/metrics/restapi/ModifiedLinesCoverageApiTest.java:
```diff
@@ -0,0 +1,61 @@
+ */
+
+class ModifiedLinesCoverageApiTest extends AbstractModifiedFilesCoverageTest {
+    //TODO: write shouldProvideRemoteApi() test
```
Yes, just forgot this

fo-code commented:
@uhafner I reworked the implementation and included your comments. Now everything works like discussed and is also tested.

The only problem is there are test failures regarding docker images and I think they are not caused by this pull request, so I do not know how to fix it.

uhafner reviewed (approved):
Thanks!

uhafner commented:
I am rerunning the checks and will merge if everything is back to green...
